src/xdsl/rewriter.py

```diff
@@ -0,0 +1,199 @@
+        return self.func(op, new_operands)
+
+
+def op_type_rewrite_pattern(func):
+    """
+    This function is intended to be used as a decorator on a RewritePatter method.
+    It uses type hints to match on a specific operation type before calling the decorated function.
+    """
+    # Get the operation argument and check that it is a subclass of Operation
+    params = [param for param in inspect.signature(func).parameters.values()]
+    if len(params) == 3:
+        has_self = True
+        expected_type = params[1].annotation
+    else:
+        has_self = False
+        expected_type = params[0].annotation
+    if not issubclass(expected_type, Operation):
+        raise Exception("op_type_rewrite_pattern expects the first non-self "
+                        "operand type hint to be an Operation subclass")
+
+    if not has_self:
+
+        def op_type_rewrite_pattern_wrapper(
+                op: Operation,
+                operands: List[SSAValue]) -> Optional[RewriteAction]:
+            if not isinstance(op, expected_type):
+                return None
+            return func(op, operands)
+
+        return op_type_rewrite_pattern_wrapper
+
+    def op_type_rewrite_pattern_wrapper(
+            self, op: Operation,
+            operands: List[SSAValue]) -> Optional[RewriteAction]:
+        if not isinstance(op, expected_type):
+            return None
+        return func(self, op, operands)
+
+    return op_type_rewrite_pattern_wrapper
+
+
+@dataclass(repr=False, eq=False)
+class OperandUpdater:
+    """
+    Provides functionality to bookkeep changed results and to access and update them.
+    """
+
+    result_mapping: Dict[SSAValue, SSAValue] = field(default_factory=dict())
+
+    def bookkeep_results(self, old_op: Operation,
+                         action: RewriteAction) -> None:
+        """Bookkeep the changes made by a rewrite action matching on `old_op`."""
+        if len(old_op.results) == 0:
+            return
+
+        assert len(old_op.results) == len(action.new_results)
+
+        for (old_res, new_res) in zip(old_op.results, action.new_results):
+            self.result_mapping[old_res] = new_res
+
+    def get_new_value(self, value: SSAValue) -> SSAValue:
+        """Get the updated value, if it exists, or returns the same one."""
+        return self.result_mapping.get(value, value)
+
+    def get_new_operands(self, op: Operation) -> [SSAValue]:
+        """Get the new operation updated operands"""
+        return [self.get_new_value(operand) for operand in op.operands]
+
+    def update_operands(self, op: Operation) -> None:
+        """Update an operation operands with the new operands."""
+        op.operands = self.get_new_operands(op)
+
+
+@dataclass(eq=False, repr=False)
+class PatternRewriteWalker:
+    """
+    Walks the IR in the block and instruction order.
+    Can walk either first the regions, or first the owner operation.
+    The walker will walk recursively on the created operation.
```

Comment:
Maybe we should mention here, that this walker will destroy it's input?

Reply:
Yes, I'll add it!